File: librarySystem/SearchEngine.cpp
#include <iostream>
#include <string>
#include "SearchEngine.h"
#include "Book.h"
#include "Video.h"
#include "Film.h"
#include "Periodical.h"
#include <vector>
#include <fstream>
#include <cstdlib>
#include <sstream>
#include <algorithm>

using namespace std;

SearchEngine::SearchEngine()
{
	string callnumber;
	string title;
	string subjects;
	string author;
	string description;
	string publisher;
	string city;
	string year;
	string series;
	string notes;

	fstream fsBook;

	//Reading the Books 
	fsBook.open("book.txt");
	if(fsBook==NULL)
		cout<<"Error while reading book.txt"<<endl;
	//Read book.txt line by line
	while(!getline(fsBook, callnumber, '|').eof() )
	{
		getline(fsBook, title,'|');
		getline(fsBook, subjects,'|');
		getline(fsBook, author,'|');
		getline(fsBook, description,'|');
		getline(fsBook, publisher,'|');
		getline(fsBook, city,'|');
		getline(fsBook, year,'|');
		getline(fsBook, series,'|');
		getline(fsBook, notes);

		Book *book = new Book(callnumber, title, subjects, author, description, publisher, city, year, series, notes);
		cardCatalog.push_back(book);
		
	}
	fsBook.close();

	//Reading the Video file
	string distributor;
	string label;
	fstream fsVideo;

	fsVideo.open("video.txt");

	if(fsVideo==NULL)
		cout<<"Error while reading video.txt"<<endl;
    //Read video.txt line by line
	while(!getline(fsVideo, callnumber, '|').eof() )
	{
		getline(fsVideo, title,'|');
		getline(fsVideo, subjects,'|');
		getline(fsVideo, description,'|');
		getline(fsVideo, distributor,'|');
		getline(fsVideo, notes,'|');
		getline(fsVideo, series,'|');
		getline(fsVideo, label);

		Video *video = new Video(callnumber, title, subjects, description, distributor, notes, series, label);
		cardCatalog.push_back(video);
		
	}
	fsVideo.close();

	//Reading the Film file
	string director;
	fstream fsFilm;
	fsFilm.open("film.txt");

	if(fsFilm==NULL)
		cout<<"Error while reading film.txt"<<endl;
	
    //Read the Film file line by line
	while(!getline(fsFilm, callnumber, '|').eof() )
	{
		getline(fsFilm, title,'|');
		getline(fsFilm, subjects,'|');
		getline(fsFilm, director,'|');
		getline(fsFilm, notes,'|');
		getline(fsFilm, year);

		Film *film = new Film(callnumber, title, subjects, director, notes, year);
		cardCatalog.push_back(film);
		
	}
	fsFilm.close();

	//Reading the Periodical file
	string publishingHistory;
	string relatedTitl;
	string otherTitles;
	string govDocNum;
	fstream fsPeriodical;
	fsPeriodical.open("periodic.txt");

	if(fsPeriodical==NULL)
		cout<<"Error while reading periodical.txt"<<endl;

	while(!getline(fsPeriodical, callnumber, '|').eof() )
	{
		getline(fsPeriodical, title,'|');
		getline(fsPeriodical, subjects,'|');
		getline(fsPeriodical, author,'|');
		getline(fsPeriodical, description,'|');
		getline(fsPeriodical, publisher,'|');
		getline(fsPeriodical, publishingHistory,'|');
		getline(fsPeriodical, series,'|');
		getline(fsPeriodical, notes,'|');
		getline(fsPeriodical, relatedTitl,'|');
		getline(fsPeriodical, otherTitles,'|');
		getline(fsPeriodical, govDocNum);

		Periodical *periodical = new Periodical(callnumber, title, subjects, author, description, publisher, publishingHistory,
			series, notes, relatedTitl, otherTitles, govDocNum);
		cardCatalog.push_back(periodical);
		
	}
	fsPeriodical.close();
}

SearchEngine::~SearchEngine()
{
    //release memory allocated while pushing in records
    for(unsigned int i=0; i < cardCatalog.size(); ++i)
    {
        delete cardCatalog[i];
    }
}

const vector<Media* > SearchEngine::searchByCallNum(const string &callNumber) const
{
	unsigned int i=0;
    vector <Media *> searchResult;
	bool found;

	for(i=0; i < cardCatalog.size(); i++)
	{
        //compare the input callnumber with every record
		found = cardCatalog[i]->compare_callNumber(callNumber);
		if(found)
		{
			searchResult.push_back(cardCatalog[i]);
		}
	}
	return searchResult;
}

const vector<Media *> SearchEngine::searchByTitle(const string &title) const
{
	unsigned int i=0;
	bool found;
    vector <Media *> searchResult;

	for(i=0; i < cardCatalog.size(); i++)
	{
		found = cardCatalog[i]->compare_title(title);
		if(found)
		{
			searchResult.push_back(cardCatalog[i]);
		}
	}
	return searchResult;
}

const vector<Media *> SearchEngine::searchBySubject(const string &subject) const
{
	unsigned int i=0;
	bool found;
    vector <Media *> searchResult;

	for(i=0; i < cardCatalog.size(); i++)
	{
		found = cardCatalog[i]->compare_subject(subject);
		if(found)
		{
			searchResult.push_back(cardCatalog[i]);
		}
	}
	return searchResult;
}

const vector<Media *> SearchEngine::searchByOther(const string &other) const
{
	unsigned int i=0;
	bool found;
    vector <Media *> searchResult;

	for(i=0; i < cardCatalog.size(); i++)
	{
		found = cardCatalog[i]->compare_other(other);
		if(found)
		{
			searchResult.push_back(cardCatalog[i]);
		}
	}
	return searchResult;
}

File: libPrjPython/SearchEngine.py
from Book import Book
from Periodical import Periodical
from Video import Video
from Film import Film

class SearchEngine(object):

    cardCatalog = []

    def __init__(self):
        book = open("book.txt", "r")
        for eachLine in book:
            bookData = eachLine.strip().split("|")
            self.cardCatalog.append(Book(bookData[0],bookData[1],bookData[2],bookData[3],bookData[4],bookData[5],bookData[6],
                                         bookData[7],bookData[8],bookData[9]))
        book.close()

        periodical = open("periodic.txt", "r")
        for eachLine in periodical:
            periodicalData = eachLine.strip().split("|")
            self.cardCatalog.append(Periodical(periodicalData[0],periodicalData[1],periodicalData[2],periodicalData[3],                                                          periodicalData[4],periodicalData[5],periodicalData[6],periodicalData[7],
                                               periodicalData[8],periodicalData[9],periodicalData[10],periodicalData[11]))
        periodical.close()

        video = open("video.txt", "r")
        for eachLine in video:
            videoData = eachLine.strip().split("|")
            self.cardCatalog.append(Video(videoData[0],videoData[1],videoData[2],videoData[3],videoData[4],videoData[5],
                                          videoData[6],videoData[7]))
        video.close()

        film = open("film.txt", "r")
        for eachLine in film:
            filmData = eachLine.strip().split("|")
            self.cardCatalog.append(Film(filmData[0],filmData[1],filmData[2],filmData[3],filmData[4],filmData[5]))
        film.close()

    
    def search_By_CallNo(self, search_str):
        result = []
        for eachItem in self.cardCatalog:
            if eachItem.compare_callno(search_str) == True:
                result.append(eachItem)
        return result

    def search_By_Title(self, search_str):
        result = []
        for eachItem in self.cardCatalog:
            if eachItem.compare_title(search_str) == True:
                result.append(eachItem)
        return result

    def search_By_Subject(self, search_str):
        result = []
        for eachItem in self.cardCatalog:
            if eachItem.compare_subject(search_str) == True:
                result.append(eachItem)
        return result

    def search_By_Other(self, search_str):
        result = []
        for eachItem in self.cardCatalog:
            if eachItem.compare_other(search_str) == True:
                result.append(eachItem)
        return result
